MVJ-403 Importing index numbers for Periodic Rent Adjustment

juho-kettunen-nc opened the pull request:
Models and an importer for indexes and index numbers necessary to perform a Periodic Rent Adjustment.

henrinie-nc commented on leasing/management/commands/import_level_check_index.py:
```diff
@@ -0,0 +1,299 @@
+        values: List of all measure column values.
+    """
+
+    key: list[str]
```
Might be slightly better to use `from typing import List` in place of `list`

juho-kettunen-nc replied:
I have understood that after Python 3.8 the non-typing generic types such as `list` are preferred, and we have now upgraded to 3.10 in production. Or do you have another reason to prefer `typing.List` besides backwards compatibility in mind?

henrinie-nc replied:
It may be that my knowledge is outdated in that case

henrinie-nc commented on leasing/management/commands/import_level_check_index.py:
```diff
@@ -0,0 +1,299 @@
+class Command(BaseCommand):
+    help = "Import indexes related to Level Check (Tasotarkistus) from Tilastokeskus database."
+
+    def handle(self, *args, **options):
```
Matter of taste, but for the handler it might be easier to understand what it does, if it was chopped into smaller named functions - ony could read the function names and understand what is happening, also it would be shorter.

juho-kettunen-nc replied:
Yes, I agree. Now that the logic works, I can extract at least two parts from the handler logic: the index itself, and the index numbers.

juho-kettunen-nc replied:
The functionality is now extracted.

henrinie-nc commented on leasing/management/commands/import_level_check_index.py:
```diff
@@ -0,0 +1,299 @@
+                index_column_pos = _find_column_position(columns, index_input["code"])
+                number_value_pos = _find_value_position(columns, index_input["code"])
+            except ResponseDataError as e:
+                self.stderr.write(f"Error: {e}. Skipping this index import")
```
For the output, would be preferable to replace `this` with the name of the index, e.g. the code

juho-kettunen-nc replied:
I assume this helps with Sentry and other error reporting tools? In that case, good idea. I did not specify the index here originally because the index name is written to `stdout` earlier in the same block.

juho-kettunen-nc replied:
Fixed.

henrinie-nc commented on leasing/management/commands/import_level_check_index.py:
```diff
@@ -0,0 +1,299 @@
+            metadata = next(iter(metadata), {})
+
+            index, created = (
+                OldDwellingsInHousingCompaniesPriceIndex.objects.update_or_create(
```
Wondering about the default values, if they don't exist is there a need to create or update anything?
The default values for `dict.get()` make it look like it would then create empty cells in data.

juho-kettunen-nc replied:
The defaults are mostly empty strings because [Django reference] (https://docs.djangoproject.com/en/4.2/ref/models/fields/#django.db.models.Field.null) advises to avoid both `null=True` and `blank=True` for CharFields and TextFields in the models, to avoid possible ambiguity in having two possibilities for "no value".

But in this project I have seen that the `null` SQL value has been the default way for representing missing string values. I can align with that, if you prefer.

henrinie-nc replied:
What I meant was that if the values are not there, should we go ahead and update existing values with empty values?

juho-kettunen-nc replied:
Ah, ok. I think my reasoning here aligns with my comment to your similar question for IndexNumber update: https://github.com/City-of-Helsinki/mvj/pull/696#discussion_r1719469218

henrinie-nc commented on leasing/management/commands/import_level_check_index.py:
```diff
@@ -0,0 +1,299 @@
+            index_column_details = columns[index_column_pos]
+
+            metadata = index_data.get("metadata", [])
+            metadata = next(iter(metadata), {})
```
What is the purpose of this?

juho-kettunen-nc replied:
`index_data["metadata"]` is a list of dictionaries, or empty list. Line 115 takes the first item in that list, or empty dictionary if not found. I wrote it this way to be concise and avoid boilerplate in handling the possibly empty list, but I can expand it to make more sense with a ternary expression or an if-else.

juho-kettunen-nc replied:
I made this more simple with a ternary expression `metadata = table_metadata[0] if table_metadata else {}`

henrinie-nc commented on leasing/management/commands/import_level_check_index.py:
```diff
@@ -0,0 +1,299 @@
+                )
+                region = dp["key"][region_key_pos]
+                comment = _find_comment_for_value(dp, comments, columns)
+                IndexNumber.objects.update_or_create(
```
Do we need any safe guards for obvious errors in the data that would cancel updating any values. For instance if the number is None and we have a previous number, do we want to overwrite the previous number? Perhaps the Model does not allow nulls, but it is not visible here without checking

juho-kettunen-nc replied:
This is something I was wondering too. This could be solved by checking the previous value before updating, but this adds more database calls for each number.

Or by supplying an illegal value as the default, so that the model rejects the update if we don't want it. This would require more error handling to catch this expected exception, so the import run doesn't stop.

This model allows for nulls, so the numbers will be overwritten if the API response would replace a previously existing number value with a period `.`. But I guess there might be some legitimate reason for Tilastokeskus to do this change, and our database should probably reflect the state that is provided by API. I don't see this as a big problem, because if the overwritten number would later receive a proper value later in the API source table, then our numbers would be overwritten again with the proper value.

juho-kettunen-nc replied:
I didn't change the null/empty value handling yet, we can discuss it more if needed.

henrinie-nc commented on leasing/models/rent.py:
```diff
@@ -1476,6 +1476,118 @@ class LegacyIndex(models.Model):
+    """
+
+    # Maximum lengths are arbitrary, but set to avoid extra large input.
+    CHARFIELD_MAX_LENGTH = 255
```
It might be that this is the default value used in Django, but does not hurt to define it so it is clear what it is

henrinie-nc commented on leasing/models/rent.py:
```diff
@@ -1476,6 +1476,118 @@ class LegacyIndex(models.Model):
+    CHARFIELD_MAX_LENGTH = 255
+
+    code = models.CharField(
+        verbose_name=_("Code"),
```
Just an idea, does the translation function take any extra arguments, e.g. describing which "code" this is for translation purposes.

juho-kettunen-nc replied:
I'll check this and implement if that is possible.

juho-kettunen-nc replied:
I see there are two ways to accomplish this:
1. Decide a unique verbose name such as `index code` instead of `code` so it generates its own localization entry
2. Use the function `pgettetxt_lazy` instead of `gettext_lazy`, so you can pass in the message context to "namespace" the localization while using generic verbose names: https://docs.djangoproject.com/en/5.0/topics/i18n/translation/#contextual-markers

I have yet to decide which I prefer, but I'll try option 1. first and then 2. if it starts making more sense.

henrinie-nc commented on leasing/models/rent.py:
```diff
@@ -1476,6 +1476,118 @@ class LegacyIndex(models.Model):
+            "model name", "price indexes of old dwellings in housing companies"
+        )
+        constraints = [
+            models.UniqueConstraint(fields=["code", "url"], name="unique_price_index")
```
Just checking: code is not unique?

juho-kettunen-nc replied:
`code` appears to be unique for the same index (in this case the `2020=100` price index for old dwellings in housing companies), but it is not unique for the resolution of the index: the same code is used for yearly and quarterly tables. These two tables have different values for other properties of that index column, because one table has the column property values in Finnish and the other in English.

For example, the source table column property `text` (which I save as the index `name` in the model) for the index column has value `Indeksi (2020=100)` in the quarterly table and `Index (2020=100)` in the yearly table. If both tables were imported, the `name` field in the index column for this `code` would churn between the two values unnecessarily. The same is true for the index column property `comment`, which is in Finnish in the quarterly table and in English in the yearly table, and which I save in the index model as `comment`.

This is not a problem right now, because we only import the quarterly table. I anticipate this churning/overwriting issue in the future, if Tilastokeskus adds more indexes in this same table for future years (2025, 2030...) and if those index columns have different values for the `text` or `comment` properties. Or if we later decide to import both the yearly and quarterly index values with this importer.

juho-kettunen-nc replied:
I thought this through once again, and the code must be unique for every index. Otherwise a single table couldn't show multiple different indexes with same "topic", such as we have now. The quarterly/yearly interval is also not a problem because we can sidestep that on our side, e.g. by leaving the quarter value as null if we want to signify a yearly index number.

--> the only unique field needs to be "code".

juho-kettunen-nc replied:
Unique constraint updated to be only `code`

henrinie-nc commented on leasing/models/rent.py:
```diff
@@ -1476,6 +1476,118 @@ class LegacyIndex(models.Model):
+    # Max lengths here are arbitrary, but set to avoid extra large input.
+    region = models.CharField(verbose_name=_("Region"), blank=True, max_length=255)
+    comment = models.TextField(verbose_name=_("Region"), blank=True)
+
```
Since you asked, I would prefer to have `updated_at` field at least. Could be useful - if there ever is a need to figure out issues. E.g. in 2035 the value for 2025Q1 is updated.

juho-kettunen-nc replied:
I have understood that the `updated_at` field will be updated every time the `update_or_create` function is called, regardless if any field values were changed or not. This would make it only and indicator when the importer was last run, and not when the field values were changed. Or does the timestamp work differently?

juho-kettunen-nc replied:
Added `updated_at` and `created_at` as part of the `TimeStampedModel`.

henrinie-nc commented:
Timestamps: halutaanko created_at, updated_at, deleted_at ?

Updated_at could be useful for the index values. Created_at in my opinion is not as important, but on the other hand it really wouldn't hurt to add both.

    Kenttien sisältö OK? Halutaanko muita?

Looks good to me, you know the data better than others :)

    Kenttien tyypit OK?

Yes

    Kenttien validointi OK?

I think so, django itself will not accept wrong kind of values for its fields.

    IndexNumber: halutaanko indeksin "code" tai teksti tähän näkyviin selaamisen helpottamiseksi? Esim "ind15" tai "Index (2015=100)"

It is in the related table in normalized for, no need to add it, one can join the tables to get the names if needed.

    Koodista ylipäänsä mielipiteitä? Tyyli, kommentit, nimeäminen, ymmärrettävyys...

Code is excellent and well written. The handler could have less of the logic itself and use named functions to describe the process and logic, but this is a matter of taste.
The data parsing part is a bit hard to understand for me at least, but I suspect it relates to the fact how the API presents the data and the format might not be the easiest.

TODO jotka korjaan vielä ennen oikeaa PR:ää:

    testejä oleellisille apufunktioille, että varmasti toimivat kuten halutaan

We could use this as an opportunity to collaboratively create tests as a learning opportunity, what do you think?

    committien uudelleen järjestely, tarpeettomat pikkumuutokset pois commiteista, jne

I would not worry about that, the feature is small enough to be squashed in one commit I think.

    migraatiotiedosto kuntoon; nyt tuosta puuttuu pari muutosta mitkä tein malliin.

It is best to migrate backwards, delete the migration file, generate a new one - instead of manually modifying it.

NC-jsAhonen reviewed (commented):
Overall, looks ok to me.

juho-kettunen-nc commented on leasing/models/rent.py:
```diff
@@ -1476,6 +1476,118 @@ class LegacyIndex(models.Model):
+        verbose_name=_("Name"),
+        max_length=CHARFIELD_MAX_LENGTH,
+    )
+    comment = models.TextField(verbose_name=_("Region"), blank=True)
```
The existing convention in DB models in backend seems to be to translate verbose name of each field. 

However, it looks like we go against the Django convention of NOT capitalizing the first letter of verbose name: The convention is not to capitalize the first letter of the verbose_name <https://docs.djangoproject.com/en/4.2/topics/db/models/#verbose-field-names>. Maybe better I align with our existing convention of capitalization instead of making an exception.

NC-jsAhonen commented on leasing/management/commands/import_periodic_rent_adjustment_index.py:
```diff
@@ -0,0 +1,407 @@
+
+            index_data = _get_index_data(index_input["url"], index_input["code"])
+            try:
+                _check_that_response_data_is_valid(index_input, index_data)
```
This could be unit tested. It is good to check the validity of the function that tests that the response data is valid :)

NC-jsAhonen commented on leasing/management/commands/import_periodic_rent_adjustment_index.py:
```diff
@@ -0,0 +1,407 @@
+        for index_input in INDEXES_TO_IMPORT:
+            self.stdout.write(f'Index: "{index_input["name"]}"')
+
+            index_data = _get_index_data(index_input["url"], index_input["code"])
```
This one accesses network, so unit testing as such is not stable. Maybe we could create a dummy data set that is representative of this function's return value, to use it to test other functions in this function.

NC-jsAhonen commented on leasing/management/commands/import_periodic_rent_adjustment_index.py:
```diff
@@ -0,0 +1,407 @@
+                )
+                continue
+
+            index, created = _update_or_create_index(index_input, index_data)
```
This function is meant to be able to both update and create. There could be a test case for both.

NC-jsAhonen commented on leasing/management/commands/import_periodic_rent_adjustment_index.py:
```diff
@@ -0,0 +1,407 @@
+                "Added new index." if created else "Updated existing index."
+            )
+
+            numbers_updated, numbers_created = _update_or_create_index_numbers(
```
The same as for `_update_or_create_index`.

juho-kettunen-nc replied:
Unit tests added

NC-jsAhonen reviewed (changes requested):
There could be unit tests for at least these functions that are used in the `import_periodic_rent_adjustment_index` command's `handle` method:
- `_check_that_response_data_is_valid`
- `_update_or_create_index`
- `_update_or_create_index_numbers`

henrinie-nc commented:
@juho-kettunen-nc Is this ready for full review (as in no longer a draft)?

juho-kettunen-nc commented:
Unit tests are in. TODO verify main feature logic after 5.9. meeting.

juho-kettunen-nc commented:
Ready for review

juho-kettunen-nc commented:
I generated the translation messages, but this time the command added source file locations for each message.

juho-kettunen-nc commented:
Converted back to draft, because fixtures are TODO

henrinie-nc commented:
> I generated the translation messages, but this time the command added source file locations for each message.

Try with management command `app_makemessages`

juho-kettunen-nc commented:
Ready for review again. I added the daily batchrun fixtures for running at 22:00 UTC. Also fixed the migrations to include the 0074 that was recently added to the develop branch.

henrinie-nc reviewed (approved):
Looks really good, great job
